storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java

```diff
@@ -507,4 +509,33 @@ public void onNewFinalizedCheckpoint(Checkpoint finalizedCheckpoint) {
         .map(ReadOnlyForkChoiceStrategy::getChainHeads)
         .orElse(Collections.emptyMap());
   }
+
+  public SafeFuture<Set<SignedBeaconBlock>> getAllBlocksAtSlot(final UInt64 slot) {
+    final Set<Bytes32> blockRoots =
+        getForkChoiceStrategy()
+            .map(forkChoiceStrategy -> forkChoiceStrategy.getBlockRootsAtSlot(slot))
+            .orElse(Collections.emptySet());
+
+    return SafeFuture.of(() -> getBlocksByBlockRoots(blockRoots));
+  }
+
+  private Set<SignedBeaconBlock> getBlocksByBlockRoots(final Set<Bytes32> blockRoots) {
+    if (blockRoots.isEmpty()) {
+      return Collections.emptySet();
+    }
+    return blockRoots.stream()
+        .map(this::retrieveSignedBlockByRoot)
+        .map(
+            future -> {
+              try {
+                return future.get();
+              } catch (Exception e) {
+                LOG.debug("Failed to get block", e);
+                final Optional<SignedBeaconBlock> res = Optional.empty();
+                return res;
+              }
+            })
```

Comment:
hmm, we can't do it this way as `.get()` is blocking which will wind up making the whole call blocking.  Need to collect all the `SafeFuture` returned from each `retrievedSignedBlockByRoot` into an array and combine with `SafeFuture.allOf` first then we can go through them all to retrieve the results (.get is safe then because we've waited for them all to complete, though `.join` will be even easier avoiding the checked exceptions).